src/sql/src/plan/query.rs

```diff
@@ -2106,13 +2107,47 @@ fn plan_view_select(
             allow_parameters: true,
             allow_windows: false,
         };
-        let expr = plan_expr(ecx, &having)
+        let expr = plan_expr(ecx, having)
             .map_err(check_ungrouped_col)?
             .type_as(ecx, &ScalarType::Bool)?;
         relation_expr = relation_expr.filter(vec![expr]);
     }
 
-    // Step 7. Handle SELECT clause.
+    // Step 7. Gather window functions from SELECT and ORDER BY, and plan them.
+    //
+    // Note that window functions can be present only in SELECT and ORDER BY (including
+    // DISTINCT ON), because they are executed after grouped aggregations and HAVING.
+    //
+    // Also note that window functions in the ORDER BY can't refer to columns introduced in the
+    // SELECT. This is because when an output column appears in ORDER BY, it can only stand alone,
+    // and can't be part of a bigger expression.
+    // See https://www.postgresql.org/docs/current/queries-order.html:
+    // "Note that an output column name has to stand alone, that is, it cannot be used in an
+    // expression"
+    let window_funcs = {
+        let mut visitor = WindowFuncCollector::default();
+        visitor.visit_select(&s);
+        for o in order_by_exprs.iter() {
+            visitor.visit_order_by_expr(o);
+        }
+        visitor.into_result()
+    };
+    for window_func in window_funcs {
+        let ecx = &ExprContext {
+            qcx,
+            name: "window function",
+            scope: &group_scope,
+            relation_type: &qcx.relation_type(&relation_expr),
+            allow_aggregates: true,
+            allow_subqueries: true,
+            allow_parameters: true,
+            allow_windows: true,
```

Comment:
`allow_windows: true` exists in 4 other places (step 4 "expand SELECT clause", step 8 "handle SELECT clause", step 9 "handle intrusive ORDER BY and DISTINCT"). Do those other 4 places need to be set to `false`? That is, should this be the exact only place where these are allowed?

Reply:
Good idea, that would add extra checks to catch if something didn't get planned here. But there are some complications:

(The one in `// Step 4. Expand SELECT clause.` should stay true I think, because it's not real planning, just doing some other stuff.)

If I set the other three to false, then I have two problems:

One problem is that if the user forgets the OVER clause from a window function, then the collector visitor won't collect it, and then we later fail with a "not allowed" error before we would print the correct error msg of missing OVER clause. I think I could fix this by modifying the collector to not look for the OVER clause, but the actual window functions.

The other problem is that if something goes wrong and we still run into an unplanned window function in one of these places, then we'll print a very confusing error msg. Something like "window functions are not allowed in SELECT clause". I'm not sure how to fix this one, maybe I should create a new flag next to `allow_windows`, something like `window_functions_should_be_already_planned`, and then fail with a custom internal error in this case. Also note that if the collector fails to collect something, then the lowering might still correctly translate the window function in most cases, so I would be hesitant to artificially make this a hard failure, when it currently might go through fine even it was not collected and planned in the newly added step.

Right now I'm running all slts with the 3 `allow_windows` set to false, to see if some other case besides the above issues pops up.

But I think we could merge it as it is now if this slt run doesn't show any more problems, and then refine it next week in a follow-up PR. I'm planning to do the switch to approach 5. from the issue description anyway in a follow-up PR next week, so then I would do the above changes in that PR. How does this sound?

Reply:
The local slt run went through without errors.

Reply:
SGTM